### ISSUE TITLE
Problematic syntax ftplugin files for folding

### ISSUE THREAD

**Issue by avidseeker** [vim/vim#16083]
### Steps to reproduce

Moving from https://github.com/neovim/neovim/issues/31233

mwe.rst:
```rst
================
Document Heading
================

Heading
=======

Sub-heading
-----------

Paragraphs are separated 
by a blank line.
```

and `vim --clean -u vimrc +'set fdm' mwe.rst` where `vimrc` is

```vim
set nocompatible
let g:rst_fold_enabled = 1      
let g:markdown_folding = 1      ":set fdm returns manual
"let g:javaScript_fold = 1       ":set fdm returns syntax
filetype plugin on
syntax on
```

Everything works as expected. Markdown folding doesn't interfere with rst
folding. However, adding any of these lines
```vim
let g:javaScript_fold = 1       ":set fdm returns syntax
let g:php_folding = 1       ":set fdm returns syntax
```

and now rst folding no longer works. I only managed to find php and javascript, but there could be other filetype plugins having the same issue, too.

### Expected behaviour

.

### Version of Vim

9.1

### Environment

Linux

**Comment by zeertzjq** [comment]
This is a problem with the syntax files, not filetype plugins.

**Comment by chrisbra** [comment]
> let g:javaScript_fold = 1       ":set fdm returns syntax
> let g:php_folding = 1       ":set fdm returns syntax

If I understand correctly, this will just enable syntax-based folding for the *embedded* javascript/php language. Why do you expect this to enable folding of your RST document?

**Comment by Konfekt** [comment]
> Why do you expect this to enable folding of your RST document?

As there were added, supposedly the line coming before  

> let g:rst_fold_enabled = 1 

entails

```vim
" /usr/local/share/vim/vim91/ftplugin/rst.vim (lines 37-44)
if g:rst_fold_enabled != 0 && has('patch-7.3.867')  " Introduced the TextChanged event.
  setlocal foldmethod=expr
  setlocal foldexpr=RstFold#GetRstFold()
  setlocal foldtext=RstFold#GetRstFoldText()
  augroup RstFold
    autocmd TextChanged,InsertLeave <buffer> unlet! b:RstFoldCache
  augroup END
endif
```

The issues stems from here

```vim
" syntax/rst.vim (line 230)
exe 'syn include @rst'.s:filetype.' syntax/'.s:filetype.'.vim'
```

since

```vim
" syntax/javascript.vim (line 82)
setlocal foldmethod=syntax
```

This is a [recurring issue](https://github.com/tpope/vim-markdown/issues/154#issuecomment-778677280) and the adhoc fix could be similar; though there should be a more generic solution for stopping embedded syntax filetypes overriding filetype local settings.

**Comment by Konfekt** [comment]
> though there should be a more generic solution for stopping embedded syntax filetypes overriding filetype local settings.

Maybe simply deferring them to `ftplugin/*.vim`, possibly to an autocmd later checking if syntax highlighting of the corresponding filetype is enabled

**Comment by chrisbra** [comment]
I think the javascript syntax file should not `set foldmethod foldtext`.  That belongs into a filetype plugin behind a configuration variable.

**Comment by clason** [comment]
Same for php. I just did a grep for `setlocal` in `runtime/syntax` with appalling results. I think a general cleanup may be called for.

**Comment by clason** [comment]
I don't think it should; the list would just have been too large to paste here.

**Comment by Konfekt** [comment]
For reference, excluding `&iskeyword`, that would include author(s), maintainer(s) or updater(s) of 

```
java.vim javascript.vim fortran.vim nroff.vim nasm.vim obse.vim hitest.vim dirpager.vim sicad.vim zsh.vim ruby.vim plsql.vim fasm.vim dylan.vim eiffel.vim qml.vim r.vim baan.vim occam.vim php.vim pf.vim upstreamrpt.vim gift.vim lpc.vim clojure.vim qf.vim redif.vim racket.vim progress.vim natural.vim lisp.vim cdl.vim 2html.vim dockerfile.vim 
```

**Comment by marshallward** [comment]
With respect to `rst.vim`, it's not clear to what action I need to take here, since I didn't implement the folding code and barely understand it myself.

It looks like an update from the the [dev repo](https://github.com/marshallward/vim-restructuredtext/) might also be required to restore folding functionality.  (This repo was set up while Bram was still with us, but perhaps it's no longer necessary.)

If someone can summarize what is needed, then I'll try to get it done.  (Or maybe the author @anntzer has the bandwidth to take this on?)

**Comment by Konfekt** [comment]
Oh, actually you're off the hook as the fault found in https://github.com/vim/vim/issues/16083#issuecomment-2485093868 is at

```vim
" syntax/javascript.vim (line 82)
setlocal foldmethod=syntax
```

and the Javascript syntax maintainer should to move it from `syntax/` to `ftplugin/` (as all those of the files listed in https://github.com/vim/vim/issues/16083#issuecomment-2485345868 ).

**Comment by Konfekt** [comment]
The maintainers who have left contact information on Github and therefore could be kindly asked to move their `set(local)` code lines from `syntax/` to `ftplugin/` would include @benknoble, @clojure-vim, @klutometis, @TysonAndre, @eiffelhub, @lee-lindley, @vim-ruby, @lacygoill (hitest), @chrisbra (ZSH), @zzzyxwvut, @honza (docker). Others can be inferred; for example, I'd suspect @fleiner to be the Github account of the `syntax/javascript.vim` maintainer

**Comment by lee-lindley** [comment]
for plsql, folding is not turned on by default. This command is only issued if the user does "let plsql_fold = 1" in their .vimrc. I'm not familiar with a requirement that this kind of thing be in ftplugin nor all that interested in changing it considering the tiny use case that is only triggered if the user opts for my easy button folding. It is a convenience for the user to turn it on this way. They are welcome to create their own ftplugin file if they choose.

**Comment by Konfekt** [comment]
> Above, all code is in one place. If setlocal isk+=$ were
to be moved to ftplugin/java.vim, I would be at a crossroads:

I think @chrisbra wanted to exclude this one due to compatibility with older Vim versions;
above file list only lists files setting other options

> The &l:foldtext setting is already an opt-in:

This was on my request https://github.com/vim/vim/commit/371bab05947b32f26d1b32922e5dc38343a875bc#commitcomment-145565642

The point remains that moving these in `ftplugin/` is less a hassle than running after every set option in `syntax/` 

> I find that rather distilling the presented
rationale in the form of a comment for why setlocals should
remain where they are in syntax/java.vim would be a far
better improvement than the suggested move.

Why not document that the user can enable sensible options, like foldtext in `ftplugin/` (rather than setting a `g:...` variable to the same effect). The `g:...` variable guard is the next best thing, but the user's setting is more flexible, by allowing for load guards such as `if &filetype = '...'`

**Comment by Konfekt** [comment]
> if the user does "let plsql_fold = 1" in their .vimrc.

That's how this issue arose for other file types with the conclusion that the check for it better be done in `ftplugin/` to avoid these

**Comment by benknoble** [comment]
> For reference, excluding `&iskeyword`, that would include author(s), maintainer(s) or updater(s) of
> 
> ```
> […] racket.vim […]
> ```

The racket syntax file only does `set syntax` in versions less than `800`, which are versions from before that syntax file was included in this repos `runtime/`. If that should be converted to `:setlocal`, [patches welcome at benknoble/vim-racket](https://github.com/benknoble/vim-racket) which will get upstreamed the next time I remember to do that.

**Comment by Konfekt** [comment]
Since changes depend on the maintainers' consent, how to keep track of them having been emailed in case no Github account was given?

**Comment by lee-lindley** [comment]
As for the plsql syntax file, the repository name is present in the comments of the syntax file. You are welcome to submit a pull request. This isn't something I have time to read enough to understand, nor do I understand why it is important. If you care enough to fix it, I'll make sure everything still works for me, then push a change and make it available to the vim maintainers.

https://github.com/lee-lindley/vim_plsql_syntax

**Comment by Konfekt** [comment]
Waiting for all concerned maintainers to react and take measures can take a while and seems less and less realisitic a perspective with up to now one pull request being merged. 

There was a similar sweeping change https://github.com/vim/vim/pull/14843 to `&commentstring` where this was carried faster. Maybe we can draw inspiration from that approach?
